[Blog] Initial a blog: a  neural sparse two-phase algorithm. (#3146)
opensearch-project/project-website: _posts/2024-08-07-Introducing-a-neural-sparse-two-phase-algorithm.md

---
layout: post
title:  Introducing the neural sparse two-phase algorithm
authors:
  - zhichaog
  - yych
  - congguan

> date: 2024-08-07

Reviewer: Please update hte publish date to today - 2024-08-13

categories:
    - technical-posts
has_science_table: true
meta_keywords: OpenSearch semantic search, neural sparse search, semantic sparse retrieval

> meta_description: Introducing a new search pipeline that can reduce the latency of OpenSearch neural sparse search.

Reviewer: Please update the meta with the following: 

_meta_keywords:_ `neural sparse search, OpenSearch semantic search, neural sparse two phase processor`

_meta_description:_ `OpenSearch continues to refine neural sparse search and retrieval with a two-phase algorithm that significantly reduces computational load while maintaining the quality of the final ranking.
`

Given these advantages, we’ve continued to refine neural sparse retrieval to make it even more efficient. OpenSearch 2.15 introduces a new feature: the two-phase search pipeline. This pipeline splits the neural sparse query terms into two categories: high-scoring tokens that are more relevant to the search and low-scoring tokens that are less relevant. Initially, the algorithm selects documents using the high-scoring tokens and then recalculates the score for those documents by including both high- and low-scoring tokens. This process significantly reduces computational load while maintaining the quality of the final ranking.

## The two-phase algorithm 

The two-phase search algorithm operates in two stages:

1. **Initial Phase:** The algorithm uses model inference to quickly select a set of candidate documents using high-scoring tokens from the query. These high-scoring tokens, which constitute a small portion of the total number of tokens, have significant weight---or relevance---allowing for a rapid identification of potentially relevant documents. This process significantly reduces the number of documents that need to be processed, thereby lowering computational costs.

> 2. **Recalculation Phase:** The algorithm then recalculates the scores for the candidate documents selected in the first phase, this time including both high-scoring and low-scoring tokens from the query. Although low-scoring tokens carry less weight individually, they provide valuable information in a comprehensive evaluation, particularly when long-tail terms contribute significantly to the overall score. This allows the algorithm to determine final document scores with greater accuracy.

Reviewer 2: ```suggestion
2. **Recalculation Phase:** The algorithm then recalculates the scores for the candidate documents selected in the first phase, this time including both high-scoring and low-scoring tokens from the query. Although low-scoring tokens carry less weight individually, they provide valuable information as part of a comprehensive evaluation, particularly when long-tail terms contribute significantly to the overall score. This allows the algorithm to determine final document scores with greater accuracy.
```

By processing documents in stages, this approach reduces computational overhead while mainitaining accuracy. The rapid selection in the first phase enhances efficiency, while the more detailed scoring in the second phase ensures accuracy. Even when handling a large number of long-tail terms, the results remain of high quality, with a notable improvement in computational efficiency.

## Performance metrics

We measured the speed and quality of search results using neural sparse search.

### Test environment

> Performance was measured on OpenSearch clusters with 3 m5.4xlarge nodes using [OpenSearch Benchmark](https://opensearch.org/docs/latest/benchmark/). The tests were conducted with 20 simultaneous clients, 50 warmup iterations, and 200 test iterations.

Reviewer 2: ```suggestion
Performance was measured on OpenSearch clusters containing 3 m5.4xlarge nodes using [OpenSearch Benchmark](https://opensearch.org/docs/latest/benchmark/). The tests were conducted with 20 simultaneous clients, 50 warmup iterations, and 200 test iterations.
```

> Performance was measured on OpenSearch clusters with 3 m5.4xlarge nodes using [OpenSearch Benchmark](https://opensearch.org/docs/latest/benchmark/). The tests were conducted with 20 simultaneous clients, 50 warmup iterations, and 200 test iterations.

Reviewer 2: Reworded

### Test dataset

For search quality, we tested multiple BEIR datasets and measured the relative quality of the results. The following table presents parameter details for these datasets.

> | Dataset        | BEIR-Name     | Queries | Corpus | Rel D/Q |

Reviewer 2: BEIR-Name is the name of the dataset.

|----------------|---------------|---------|--------|---------|
| NQ             | nq            | 3,452   | 2.68M  | 1.2     |
| HotpotQA       | hotpotqa      | 7,405   | 5.23M  | 2       |
| DBPedia        | dbpedia-entity| 400     | 4.63M  | 38.2    |
| FEVER          | fever         | 6,666   | 5.42M  | 1.2     |
| Climate-FEVER  | climate-fever | 1,535   | 5.42M  | 3       |

### p99 latency

> 

Reviewer 3: Line below, first sentence: This reads as though a word should follow "search". "operation"?

Reviewer 2: Reworded.

> The algorithm maintains the same inference time cost as the existing neural sparse search. To provide a clearer comparison of acceleration in the search phase, we excluded the inference step from latency calculations because inference is significantly affected by hardware type. The latency benchmark provided in this post uses raw vector search and excludes any additional impact from inference time.

Reviewer 3: ```suggestion
The algorithm maintains the same inference time cost as the existing neural sparse search. To provide a clearer comparison of acceleration in the search phase, we excluded the inference step from latency calculations because inference is significantly affected by hardware type. The latency benchmark provided in this post uses raw vector search and excludes any additional impact resulting from inference time.
```

#### Doc-only mode

In doc-only mode, the two-phase processor can significantly decrease query latency, as shown in the following figure.

![Two-Phase Doc Model P99 Latency](/assets/media/blog-images/2024-08-07-Introducing-a-neural-sparse-two-phase-algorithm/two-phase-doc-model-p99-latency.jpg)

> The **average latency** is:

Reviewer 3: ```suggestion
The following is the **average latency**:
```

> * Without two-phase algorithm: **198 ms**

Reviewer 3: ```suggestion
* Without the two-phase algorithm: **198 ms**
```

> * With two-phase algorithm: **124 ms**

Reviewer 3: ```suggestion
* With the two-phase algorithm: **124 ms**
```

> Depending on the data distribution, the two-phase processor achieved a **speedup ranging from 1.22x to 1.78x**.

Reviewer 3: ```suggestion
Depending on the data distribution, the two-phase processor achieved an **increase in speed ranging from 1.22x to 1.78x**.
```

#### Bi-encoder mode

In bi-encoder mode, the two-phase algorithm can significantly decrease query latency, as shown in the following figure.

![Two-Phase Bi-Encoder P99 Latency](/assets/media/blog-images/2024-08-07-Introducing-a-neural-sparse-two-phase-algorithm/two-phase-bi-encoder-p99-latency.jpg)

> The **average latency** is:

Reviewer 3: ```suggestion
The following is the **average latency**:
```

> * Without two-phase algorithm: **617 ms**

Reviewer 3: ```suggestion
* Without the two-phase algorithm: **617 ms**
```

> * With two-phase algorithm: **122 ms**

Reviewer 3: ```suggestion
* With the two-phase algorithm: **122 ms**
```

> Depending on the data distribution, the two-phase processor achieved a **speedup ranging from 4.15x to 6.87x**.

Reviewer 3: ```suggestion
Depending on the data distribution, the two-phase processor achieved an **increase in speed ranging from 4.15x to 6.87x**.
```

## Try it out

To try the two-phase processor, follow these steps.

### Step 1: Set up a `neural_sparse_two_phase_processor`

First, configure a `neural_sparse_two_phase_processor` with the default parameters:

```json
PUT /_search/pipeline/<custom-pipeline-name>
{
  "request_processors": [
    {
      "neural_sparse_two_phase_processor": {
        "tag": "neural-sparse",
        "description": "This processor creates a neural sparse two-phase processor, which can speed up neural sparse queries!"
      }
    }
  ]
}
```

### Step 2: Set the default search pipeline to `neural_sparse_two_phase_processor`

> Assuming you already have a neural sparse index, set the index’s `index.search.default_pipeline` to the pipeline created in the previous step:

Reviewer 3: ```suggestion
Assuming that you already have a neural sparse index, set the index's `index.search.default_pipeline` to the pipeline created in the previous step:
```
